### PR TITLE
SOTI-3997 - Grid based interest management (p1)

### PR DESCRIPTION
**Pull request by Samstiles**

### CODE REVIEW

**Review by x-goose-x — APPROVED** [review]

File: core/instance/GridBasedInterestManagement.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
+import EDictionary from "../../external/EDictionary";
+
+class GridCell {
+    constructor() {
+        this.numberOfStaticEntities = 0;
```
Comment (MikeSchurman) [inline core/instance/GridBasedInterestManagement.js:5]: ahh js, training wheels are off now :D

File: core/instance/GridBasedInterestManagement.js [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        } else {
+            const cellToBeAddedTo = this.getCellFromWorldPos(entity.x, entity.y)
+
+            entity.nengiInternalSpatialGridCell = cellToBeAddedTo
+
+            let indexInGridCellEntityList
+
+            if (entity.isStaticEntity) {
+                indexInGridCellEntityList = cellToBeAddedTo.staticEntities.push(entity) - 1
+                cellToBeAddedTo.numberOfStaticEntities += 1;
```
Comment (MikeSchurman) [inline core/instance/GridBasedInterestManagement.js:72]: would this always be equal to `cellsToBeAddedTo.staticEntities.length`? Is this an opt?

Comment (Samstiles) [reply core/instance/GridBasedInterestManagement.js:72]: yes this is an opt, "caching" the current length of all of these lists every time the length changes

File: core/instance/GridBasedInterestManagement.js [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        // this.numberOfEntitiesNeedingToBeMovedToANewCellThisTick = 0
+        for (let i = 0; i < this.numberOfEntitiesInMasterList; i++) {
+
+            let entity = this.masterEntityList[i]
+
+            // static entities do not move and will never need to have their cell updated
+            if (entity.isStaticEntity) continue
+
+            // Big entities dont exist in cells, skip
+            if (entity.visibilityRadius !== undefined && entity.visibilityRadius >= this.thresholdToBeConsideredABigEntity) continue
```
Comment (MikeSchurman) [inline core/instance/GridBasedInterestManagement.js:246]: future opt: could probably do both of these checks not during the update (ie don't put static or large entities in the list that update iterates over)

File: index.d.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ interface EDictionary {
 
 declare namespace nengi {
     export class Instance {
-        constructor(config: any, webConfig: any)
+        constructor(config: any, webConfig: any, worldWidth: number, worldHeight: number, gridCellSize: number)
```
Comment (MikeSchurman) [inline index.d.ts:38]: unused param?

**Review by MikeSchurman — APPROVED** [review]

**Pull request merged**